File: Step 06 LinkedList/Step 6.2 Doubly Linked List/Delete node in Doubly Linked List/Delete node in Doubly Linked List.java


//User function Template for Java

/* Structure of linkedlist node
class Node
{
	int data;
	Node next;
	Node prev;
	Node(int d)
	{
		data = d;
		next = prev = null;
	}
}
*/
class Solution
{
    // function returns the head of the linkedlist
    Node deleteNode(Node head,int x)
    {
	    // Your code here
	    if(x==1){
	        head.next.prev = null;
	        head = head.next;
	        return head;
	    }
	    Node temp = head;
	    int i = 1;
	    while( i<x-1 ){
	        temp = temp.next;
	        i++;
	    }
	    if(temp.next.next != null ) temp.next.next.prev = temp;
	    temp.next = temp.next.next;
	    return head;
    }
}

File: Step 06 LinkedList/Step 6.2 Doubly Linked List/Delete node in Doubly Linked List/Delete node in Doubly Linked List.cpp
/* Structure of Node
struct Node
{
  int data;
  struct Node *next;
  struct Node *prev;

  Node(int x){
      data = x;
      next = NULL;
      prev = NULL;
  }

};
*/

class Solution {
  public:
    Node* deleteNode(Node* head, int x) {
        // Your code here
         int count = 1;
        Node* temp = head;

        while(count != x){
            temp = temp -> next;
            count++;
        }
        
        if(temp -> prev == NULL){
            temp = temp -> next;
            head = temp;
            temp -> prev = NULL;
        }
        
        else if(temp -> next == NULL){
            Node* nd = temp -> prev;
            nd -> next = NULL;
            temp -> prev = NULL;
        }
        
        else{
            Node* pr = temp -> prev;
            Node* nx = temp -> next;
            
            temp -> next = pr -> next -> next;
            temp -> prev = nx -> prev -> prev;
  
            temp -> next -> prev = temp -> prev;
            temp -> prev -> next = temp -> next;
            
            temp -> next == NULL;
            temp -> prev == NULL;
  
        }
        return head;
    }
};